### PR TITLE
Replace rest-client with Faraday

### PR DESCRIPTION
**Pull request by samjhoward**
Move most network requests to Faraday as rest-client has issues on Win

### CODE REVIEW

File: lib/buildvu.rb [hunk 1]
```diff
@@ -97,30 +99,46 @@ def download_result(results, output_file_path, file_name=nil)
...
+        f.request :multipart
+        f.request :url_encoded
+        f.adapter Faraday.default_adapter
+      end
+
+	  r = conn.post(uri.path, params)
+		
+	rescue StandardError => e
       raise("Error sending url:\n" + e.to_s)
     end
```
Comment (elepedus) [inline lib/buildvu.rb]: Can we fix this indentation? I think we may have a stray 'end' on line 118 (a leftover from the begin/rescue block?), but the indentation makes the fragment quite awkward to read, so I might be wrong.

File: lib/buildvu.rb [hunk 2]
```diff
@@ -97,30 +99,46 @@ def download_result(results, output_file_path, file_name=nil)
...
+	
+    r.status == 200 ? uuid = JSON.parse(r.body)['uuid'] : raise("Error uploading file:\n Server returned response\n" + r.status.to_s + " - " + JSON.parse(r.body)['error'])
 
     uuid.nil? ? raise("Error uploading file:\nServer returned null UUID") : uuid
   end
 
   # Check conversion status
   def poll_status(uuid)
+    uri = URI(@endpoint)
+    host = uri.scheme + "://" + uri.host + ':' + uri.port.to_s;
```
Comment (elepedus) [inline lib/buildvu.rb]: Why is this needed? Can we not pass the URI directly to Faraday?

File: lib/buildvu.rb [hunk 3]
```diff
@@ -135,7 +153,7 @@ def download(download_url, output_file_path)
       }
       RestClient::Request.new(method: :get, url: download_url, block_response: block).execute
```
Comment (elepedus) [inline lib/buildvu.rb]: Is there a reason we're still using RestClient here, when we've moved to Faraday everywhere else?

File: lib/buildvu.rb [hunk 4]
```diff
@@ -20,8 +20,10 @@
 # License:: Apache 2.0
 
 require 'json'
+require 'Faraday'
```
Comment (RFoley) [inline lib/buildvu.rb]: Don't forget to add Faraday to the gemspec! :)

File: lib/buildvu.rb (hunk 3 above)
Comment (samjhoward) [reply lib/buildvu.rb]: Only because Faraday has so much boilerplate and I believe the rest-client issue won't come into play here. I can swap it over if you feel strongly about it!

File: lib/buildvu.rb (hunk 2 above)
Comment (samjhoward) [reply lib/buildvu.rb]: Annoyingly, Faraday's connection objects seems to ignore everything after the port, so for clarity's sake I've done that before we pass it in, since we need to have the uri so we can pass uri.path into conn.post() later on. So technically we could use @endpoint instead of host, but either way we need the URI.

File: lib/buildvu.rb (hunk 1 above)
Comment (samjhoward) [reply lib/buildvu.rb]: Urgh apologies, Idea messed this one up for me...

The end on line 118 goes with the begin on line 107. The end on line 112 goes with the do on line 108... I'll fix.

File: lib/buildvu.rb (hunk 4 above)
Comment (samjhoward) [reply lib/buildvu.rb]: Oh hey Rob! I will totally do that thing that I didn't know I needed to do. >_>

File: lib/buildvu.rb (hunk 1 above)
Comment (elepedus) [reply lib/buildvu.rb]: Ah, yes, of course. My brain ignored the 'do' and paired it up with the end on 112 🙃

Comment (samjhoward) [reply lib/buildvu.rb]: Should look better if you show all commits now. :)

File: lib/buildvu.rb (hunk 2 above)
Comment (elepedus) [reply lib/buildvu.rb]: How about something like
```ruby
  def poll_status(uuid)
    uri = URI(@endpoint)
    # No need to build the host string here
    begin
      conn = Faraday.new(uri) do |req|        # just pass the uri object in here
          req.params['uuid'] = uuid
          req.request :url_encoded
		  req.adapter Faraday.default_adapter
	  end
	  r = conn.get(uri.path)        # we can still grab the path from it here
    rescue StandardError => e
      raise("Error checking conversion status:\n" + e.to_s)
    end

    r.status == 200 ? response = JSON.parse(r.body) : raise("Error checking conversion status:\n Server returned response\n" + r.status.to_s + " - " + JSON.parse(r.body)['error'])

    response

```

Comment (samjhoward) [reply lib/buildvu.rb]: Slight preference to keep it as is - Faraday could change in future to stop stripping out path, in which case we'd have double path.

File: lib/buildvu.rb (hunk 3 above)
Comment (elepedus) [reply lib/buildvu.rb]: I'm not a fan of having two different libraries for the same thing in a single project, since it introduces dependency bloat, and a cognitive overhead of two different approaches to doing the same thing. I would prefer for all connection code to use a single library, but check with Leon/Kieran before letting me make work for you! 😅

File: lib/buildvu.rb (hunk 4 above)
Comment (RFoley) [reply lib/buildvu.rb]: Hey 👋🏻😁

These should help:
[Gemspec docs: add_runtime_dependency](https://guides.rubygems.org/specification-reference/#add_runtime_dependency) // [Gem version patterns](https://guides.rubygems.org/patterns/#pessimistic-version-constraint)

Just drop me a message if you need a hand!

File: lib/buildvu.rb (hunk 3 above)
Comment (RFoley) [reply lib/buildvu.rb]: +1 to using the same lib everywhere if it's not too much hassle!

Also not trying to create additional work for you - but if Faraday isn't the nicest fit, I've been using [HTTParty](https://github.com/jnunemaker/httparty) for my most recent stuff. Their code for [streaming a file](https://github.com/jnunemaker/httparty/blob/master/examples/stream_download.rb) looks a lot nicer that my old rest-client implementation, at the very least... 😅😉

**Comment by RFoley** [comment]
Out of curiosity, what were the issues with rest-client?

**Comment by samjhoward** [comment]
> Out of curiosity, what were the issues with rest-client?
Takes down the whole app silently when called from a thread. https://github.com/rest-client/rest-client/issues/706

At the time of our discovery of this, the last commit was in 2017 and the project admins were completely MIA... looks like that might have changed now.

File: lib/buildvu.rb (hunk 4 above)
Comment (samjhoward) [reply lib/buildvu.rb]: https://github.com/idrsolutions/buildvu-ruby-client/pull/11/commits/d22aca67adac84aa4d547c47eab9b3df526821aa

Think that should do it!

File: lib/buildvu.rb (hunk 3 above)
Comment (samjhoward) [reply lib/buildvu.rb]: Replaced the rest with Faraday too: https://github.com/idrsolutions/buildvu-ruby-client/pull/11/commits/5e6a7eee933ad8561c13c84cbf5b661d40d58c5f

Possibly not as nice stream-wise as before, but nice & concise. I'm keen not to spend much more time on Ruby as it's the least used of all of the clients and (as per usual) I have font issues up the wazoo!

Comment (elepedus) [reply lib/buildvu.rb]: Looks good to me -- thanks :)
